include/veriblock/blockchain/blocktree.hpp

```diff
@@ -148,15 +169,74 @@ struct BlockTree {
       current->height = current->pprev->height + 1;
       current->chainWork = current->pprev->chainWork + getBlockProof(block);
     } else {
-      current->height = 0;
       current->chainWork = getBlockProof(block);
     }
-
     determineBestChain(activeChain_, *current);
 
     return current;
   }
 
+  void invalidateBlockFromChain(Chain<Block>& chain, index_t** block) {
+    if (*block == nullptr) {
+      return;
+    }
+
+    if (!chain.contains(*block) &&
+        chain.tip()->getAncestor((*block)->height) != *block) {
+      return;
+    }
+
+    while (chain.tip() != nullptr && chain.tip() != *block) {
+      disconnectTipFromChain(chain);
+    }
+
+    if (chain.tip() != nullptr) {
+      disconnectTipFromChain(chain);
+      *block = nullptr;
+    }
+  }
+
+  void disconnectTipFromChain(Chain<Block>& chain) {
+    BlockIndex<Block>* currentTip = chain.tip();
+    hash_t tipHash = currentTip->getHash();
+
+    chain.disconnectTip();
+    block_index_.erase(tipHash.template trimLE<prev_block_hash_t::size()>());
+  }
+
+  void addForkCandidate(BlockIndex<Block>* newCandidate,
+                        BlockIndex<Block>* oldCandidate) {
+    if (newCandidate == nullptr) {
+      return;
+    }
+
+    bool isAdded = false;
+    for (auto chain_it = fork_chains_.begin();
+         chain_it != fork_chains_.end();) {
+      if (chain_it->tip() == newCandidate->pprev) {
+        chain_it->setTip(newCandidate);
+        isAdded = true;
+      }
+
+      if (chain_it->tip() == newCandidate) {
+        isAdded = true;
+      }
+
+      if (chain_it->tip() == oldCandidate) {
+        chain_it = fork_chains_.erase(chain_it);
+        continue;
+      }
+      ++chain_it;
+    }
```

Comment:
It seems there are two not connected actions: deleting oldCandidate and adding newCandidate. I think they should reside in two different functions.